src/lib/storage/index/partial_hash/partial_hash_index_impl.hpp

```diff
@@ -0,0 +1,150 @@
+#pragma once
+
+#include <tsl/sparse_map.h>
+
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "all_type_variant.hpp"
+#include "storage/chunk.hpp"
+#include "storage/index/abstract_table_index.hpp"
+#include "types.hpp"
+
+namespace hyrise {
+
+class PartialHashIndexTest;
+
+/**
+ * Forward iterator that iterates over a tsl::sparse_map that maps a DataType to a vector of RowIDs. The
+ * iteration process is as if the map would have been flattened and then been iterated.
+ *
+ * @tparam DataType The key type of the underlying map.
+ */
+template <typename DataType>
+class TableIndexMapIterator : public AbstractTableIndexIterator {
+ public:
+  using MapIterator = typename tsl::sparse_map<DataType, std::vector<RowID>>::const_iterator;
+
+  explicit TableIndexMapIterator(MapIterator it);
+
+  reference operator*() const final;
+
+  TableIndexMapIterator& operator++() final;
+
+  bool operator==(const AbstractTableIndexIterator& other) const final;
+
+  bool operator!=(const AbstractTableIndexIterator& other) const final;
+
+  std::shared_ptr<AbstractTableIndexIterator> clone() const final;
+
+  // Creates an instance of TableIndexMapIterator using the passed parameter and wraps that instance in an
+  // IteratorWrapper. Return that wrapper.
+  static IteratorWrapper create_wrapper(MapIterator it);
+
+ private:
+  MapIterator _map_iterator;
+  size_t _vector_index;
+};
+
+/**
+ * Forward iterator that iterates over a std::vector of RowIDs.
+ */
+class TableIndexVectorIterator : public AbstractTableIndexIterator {
+ public:
+  using VectorIterator = typename std::vector<RowID>::const_iterator;
+
+  explicit TableIndexVectorIterator(VectorIterator it);
+
+  reference operator*() const final;
+
+  TableIndexVectorIterator& operator++() final;
+
+  bool operator==(const AbstractTableIndexIterator& other) const final;
+
+  bool operator!=(const AbstractTableIndexIterator& other) const final;
+
+  std::shared_ptr<AbstractTableIndexIterator> clone() const final;
+
+  // Creates an instance of TableIndexVectorIterator using the passed parameter and wraps that instance in an
+  // IteratorWrapper. Return that wrapper.
+  static IteratorWrapper create_wrapper(VectorIterator it);
+
+ private:
+  VectorIterator _vector_iterator;
+};
+
+/**
+ * Base class that holds a PartialHashIndexImpl object with the correct resolved datatype.
+ */
+class BasePartialHashIndexImpl : public Noncopyable {
+  friend PartialHashIndexTest;
+
+ public:
+  using Iterator = IteratorWrapper;
+  using IteratorPair = std::pair<Iterator, Iterator>;
```

Comment:
Why not stick with `AbstractTableIndex::Iterator`? To make more understandable what is used here?